### PR TITLE
[REVIEW] Java and Jni support for Struct columns [skip ci]

### PR DESCRIPTION
**Pull request by kuhushukla**
Adds support for structs on java and jni side.

Test API and some other utils can be consolidated and I will see how we can make it better but this PR is open for comments/reviews.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 1]
```diff
@@ -2985,7 +2996,7 @@ public long getColumnViewAddress() {
 
   @Override
   public ColumnViewAccess getChildColumnViewAccess(int childIndex) {
-    if (getDataType() != DType.LIST) {
+    if (getDataType() != DType.LIST && getDataType() != DType.STRUCT) {
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/ColumnVector.java:2999]: Why not used `isNextedType()` here too?

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 2]
```diff
@@ -3024,12 +3035,16 @@ public long getNumRows() {
...
   public int getNumChildren() {
-    if (getDataType() != DType.LIST /* or structs */) {
+    if (getDataType() != DType.LIST && getDataType() != DType.STRUCT) {
       return 0;
     }
     return offHeap.getNumChildren(getNativeView());
   }
 
+  private boolean isNestedType() {
+    return type == DType.LIST || type == DType.STRUCT;
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/ColumnVector.java:3045]: I would prefer to have the DType indicate if it is a nested type or not, so it is more maintainable.

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java [hunk 3]
```diff
@@ -198,7 +198,7 @@ public long getHostMemorySize() {
     }
   }
 
-  private final OffHeapState offHeap;
+  final OffHeapState offHeap;
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/HostColumnVector.java:201]: Who needs direct access to another columns off heap?  That feels like we are exposing too much of the internals and increasing coupling.

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java [hunk 4]
```diff
@@ -455,6 +455,32 @@ public ColumnVector copyToDevice() {
...
   /////////////////////////////////////////////////////////////////////////////
 
+  /**
+   * WARNING: Strictly for test only. This call is not efficient for production.
+   */
+  ColumnBuilder.StructData getStruct(long rowIndex, ColumnBuilder.DataType mainType) {
+    assert rowIndex < rows;
+    assert type == DType.STRUCT;
+    List<Object> retList = new ArrayList<>();
+    int start = offHeap.offsets.getInt(rowIndex * DType.INT32.getSizeInBytes());
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/HostColumnVector.java:465]: Is a struct and a list some how combined in cudf??? @nvdbaranec 

https://arrow.apache.org/docs/format/Columnar.html#struct-layout shows no offsets in a struct at all.

**Review by revans2 — COMMENTED** [review]
I didn't finish looking at everything, but I am more than a little concerned about how structs are showen with an offset.

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 5]
```diff
@@ -180,12 +181,20 @@ public ColumnVector(DType type, long rows, Optional<Long> nullCount,
 
   public ColumnVector(DType type, long rows, Optional<Long> nullCount,
                       DeviceMemoryBuffer dataBuffer, DeviceMemoryBuffer validityBuffer,
-                      DeviceMemoryBuffer offsetBuffer, NestedColumnVector nestedColumnVector) {
-    if (type != DType.STRING && type != DType.LIST) {
-      assert offsetBuffer == null : "offsets are only supported for STRING, LISTS";
+                      DeviceMemoryBuffer offsetBuffer, List<NestedColumnVector> nestedColumnVectors) {
+    if (type != DType.STRING && type != DType.LIST && type != DType.STRUCT) {
```
[the remark this answers is not in the document]
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/ColumnVector.java:185]: Let me look into that. Thanks for pointing this out. I might have missed this detail.

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java (hunk 3 above)
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/HostColumnVector.java:201]: Artifact of testing . will remove.

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java (hunk 2 above)
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/ColumnVector.java:3045]: Sure, I like that, let me fix that up

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java [hunk 6]
```diff
@@ -455,6 +455,32 @@ public ColumnVector copyToDevice() {
...
+    int end = offHeap.offsets.getInt((rowIndex + 1) * DType.INT32.getSizeInBytes());
+    // check if null or empty
+    if (start == end) {
+      System.out.println("MAYBE RETURNING NULL");
```
Comment (kuhushukla) [inline java/src/main/java/ai/rapids/cudf/HostColumnVector.java:469]: Fix this debug!

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/ColumnVector.java:185]: I think it is the way I am making the cudf column view and then getting it back with an offset tht may be the issue. I  am testing this postulate right now and will update here

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java (hunk 3 above)
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/HostColumnVector.java]: Done.

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java (hunk 2 above)
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/ColumnVector.java]: Done.

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java (hunk 1 above)
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/ColumnVector.java]: Added in this and other places

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/ColumnVector.java]: @revans2 Fixed this and test pass. Could use your reviews on if things are looking ok/better and if I missed anything.

File: java/src/main/native/src/ColumnVectorJni.cpp [hunk 7]
```diff
@@ -1246,12 +1248,23 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_makeCudfColumnView(
...
+      cudf::size_type *offsets = reinterpret_cast<cudf::size_type *>(j_offset);
+      cudf::column_view offsets_column(cudf::data_type{cudf::type_id::INT32}, size + 1, offsets);
+      ret.reset(new cudf::column_view(cudf::data_type{cudf::type_id::LIST}, size, nullptr, valid,
                                                  j_null_count, 0, {offsets_column, *child_view}));
+   } else if (n_type == cudf::type_id::STRUCT) {
+     cudf::jni::native_jpointerArray<long> children(env, j_children);
+     jlong *childs = env->GetLongArrayElements(j_children, 0);
+     std::vector<column_view> children_vector(children.size());
+     for (int i = 0; i < children.size(); i++) {
+       cudf::column_view *child_view = reinterpret_cast<cudf::column_view *>(childs[i]);
```
Comment (kuhushukla) [inline java/src/main/native/src/ColumnVectorJni.cpp:1263]: Open to ideas on whether there is a better way to do this.

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java [hunk 8]
```diff
@@ -1196,7 +1225,7 @@ public static HostColumnVector timestampNanoSecondsFromBoxedLongs(Long... values
   /**
    * WARNING: Debug only method to print a passed in buffer
    */
-  private static void printBuffer(HostMemoryBuffer buffer) {
+  static void printBuffer(HostMemoryBuffer buffer) {
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/HostColumnVector.java:1228]: Does this need to be public in the final version?

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java [hunk 9]
```diff
@@ -1276,12 +1305,20 @@ public ColumnBuilder appendLists(List... inputLists) {
       return this;
     }
 
-    private void initAndResizeOffsetBuffer() {
+    public ColumnBuilder appendStructValues(List<StructData>... inputLists) {
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/HostColumnVector.java:1308]: Why an array of lists of structs?

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java [hunk 10]
```diff
@@ -1312,50 +1349,77 @@ public final ColumnBuilder appendNull() {
...
+    private <T> ColumnBuilder append(List<T> inputList) {
+      assert type.isNestedType();
+      if (type == DType.STRUCT) {
+        assert !inputList.isEmpty();
+        //TODO: assumption
+        int numChildren = inputList.size();
+        for (int i = 0; i < numChildren; i++) {
+          StructData structData = (StructData) inputList.get(i);
+          if (structData.dataRecord == null) {
+            setNullAt(currentIndex);
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/HostColumnVector.java:1368]: One of the things we have to be sure we do is that when we set a null at the parent struct level we reflect that null in the children as well.  This is one of the assumptions that we made in the cudf side to make pulling data out of a struct cheap.

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 11]
```diff
@@ -168,7 +168,8 @@ public ColumnVector(DType type, long rows, Optional<Long> nullCount,
       assert offsetBuffer == null : "offsets are only supported for STRING";
     }
 
-    offHeap = new OffHeapState(type, (int) rows, nullCount, dataBuffer, validityBuffer, offsetBuffer, null, 0l);
+    long[] children = new long[] {0l};
```
[the remark this answers is not in the document]
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/ColumnVector.java:171]: Agree.

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java (hunk 10 above)
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/HostColumnVector.java:1368]: I will add a test to verify this.

File: java/src/main/native/src/ColumnVectorJni.cpp [hunk 12]
```diff
@@ -1246,12 +1248,23 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_makeCudfColumnView(
...
+      JNI_NULL_CHECK(env, j_offset, "offset is null", 0);
+      jlong *childs = env->GetLongArrayElements(j_children, 0);
+      cudf::column_view *child_view = reinterpret_cast<cudf::column_view *>(childs[0]);
+      cudf::size_type *offsets = reinterpret_cast<cudf::size_type *>(j_offset);
+      cudf::column_view offsets_column(cudf::data_type{cudf::type_id::INT32}, size + 1, offsets);
+      ret.reset(new cudf::column_view(cudf::data_type{cudf::type_id::LIST}, size, nullptr, valid,
                                                  j_null_count, 0, {offsets_column, *child_view}));
+   } else if (n_type == cudf::type_id::STRUCT) {
+     cudf::jni::native_jpointerArray<long> children(env, j_children);
+     jlong *childs = env->GetLongArrayElements(j_children, 0);
```
[the remark this answers is not in the document]
Comment (kuhushukla) [reply java/src/main/native/src/ColumnVectorJni.cpp:1260]: Thanks for the pointer here, I did have a comment for myself and reviewers to suggest something that would be more concise.

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java (hunk 8 above)
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/HostColumnVector.java:1228]: Yes, making it so will be useful.

Comment (revans2) [reply java/src/main/java/ai/rapids/cudf/HostColumnVector.java:1228]: The like I said when this was added originally if this is going to be a public API it should move to HostMemoryBuffer.

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java (hunk 10 above)
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/HostColumnVector.java]: @revans2  The latest changes do not address this one comment. I will cover it now in the next set of changes.

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java (hunk 9 above)
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/HostColumnVector.java]: Unnecessary and fixed. Thanks!

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java (hunk 8 above)
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/HostColumnVector.java]: Done.

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 13]
```diff
@@ -3397,30 +3414,37 @@ private static NestedColumnVector createNewNestedColumnVector(
...
-        createNewNestedColumnVector(nestedChildren.getNestedChildren().isEmpty() ?
-          null : nestedChildren.getNestedChildren().get(0)));
+        children);
     }
 
     private long getViewHandle() {
-      long childColumnViewHandle = 0l;
-      if (child != null) {
-        childColumnViewHandle = child.getViewHandle();
+      long[] childrenColViews = new long[] {0L};
```
[the remark this answers is not in the document]
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/ColumnVector.java]: Done.

File: java/src/main/native/src/ColumnVectorJni.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (kuhushukla) [reply java/src/main/native/src/ColumnVectorJni.cpp]: Done.

File: java/src/main/native/src/ColumnVectorJni.cpp [hunk 14]
```diff
@@ -1246,12 +1248,23 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_makeCudfColumnView(
...
-         ret.reset(new cudf::column_view(cudf::data_type{cudf::type_id::LIST}, size, nullptr, valid,
+      JNI_NULL_CHECK(env, j_offset, "offset is null", 0);
+      jlong *childs = env->GetLongArrayElements(j_children, 0);
+      cudf::column_view *child_view = reinterpret_cast<cudf::column_view *>(childs[0]);
+      cudf::size_type *offsets = reinterpret_cast<cudf::size_type *>(j_offset);
+      cudf::column_view offsets_column(cudf::data_type{cudf::type_id::INT32}, size + 1, offsets);
+      ret.reset(new cudf::column_view(cudf::data_type{cudf::type_id::LIST}, size, nullptr, valid,
                                                  j_null_count, 0, {offsets_column, *child_view}));
+   } else if (n_type == cudf::type_id::STRUCT) {
+     cudf::jni::native_jpointerArray<long> children(env, j_children);
```
[the remark this answers is not in the document]
Comment (kuhushukla) [reply java/src/main/native/src/ColumnVectorJni.cpp]: Done.

File: java/src/main/native/src/ColumnVectorJni.cpp [hunk 15]
```diff
@@ -1246,12 +1248,23 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_makeCudfColumnView(
...
                                       j_null_count, 0, {offsets_column, data_column}));
     } else if (n_type == cudf::type_id::LIST) {
-         JNI_NULL_CHECK(env, j_offset, "offset is null", 0);
-         cudf::column_view *child_view = reinterpret_cast<cudf::column_view *>(j_child);
-         cudf::size_type *offsets = reinterpret_cast<cudf::size_type *>(j_offset);
-         cudf::column_view offsets_column(cudf::data_type{cudf::type_id::INT32}, size + 1, offsets);
-         ret.reset(new cudf::column_view(cudf::data_type{cudf::type_id::LIST}, size, nullptr, valid,
+      JNI_NULL_CHECK(env, j_offset, "offset is null", 0);
+      jlong *childs = env->GetLongArrayElements(j_children, 0);
+      cudf::column_view *child_view = reinterpret_cast<cudf::column_view *>(childs[0]);
```
[the remark this answers is not in the document]
Comment (kuhushukla) [reply java/src/main/native/src/ColumnVectorJni.cpp]: Pending.

File: java/src/main/native/src/ColumnVectorJni.cpp [hunk 16]
```diff
@@ -1246,12 +1248,23 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_makeCudfColumnView(
       ret.reset(new cudf::column_view(cudf::data_type{cudf::type_id::STRING}, size, nullptr, valid,
                                       j_null_count, 0, {offsets_column, data_column}));
     } else if (n_type == cudf::type_id::LIST) {
-         JNI_NULL_CHECK(env, j_offset, "offset is null", 0);
-         cudf::column_view *child_view = reinterpret_cast<cudf::column_view *>(j_child);
-         cudf::size_type *offsets = reinterpret_cast<cudf::size_type *>(j_offset);
-         cudf::column_view offsets_column(cudf::data_type{cudf::type_id::INT32}, size + 1, offsets);
-         ret.reset(new cudf::column_view(cudf::data_type{cudf::type_id::LIST}, size, nullptr, valid,
+      JNI_NULL_CHECK(env, j_offset, "offset is null", 0);
+      jlong *childs = env->GetLongArrayElements(j_children, 0);
```
[the remark this answers is not in the document]
Comment (kuhushukla) [reply java/src/main/native/src/ColumnVectorJni.cpp]: Done.

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/ColumnVector.java]: Done.

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 17]
```diff
@@ -3375,14 +3389,17 @@ static ColumnVector createColumnVector(DType type, int rows,
...
         mainValidDevBuff.copyFromHostBuffer(mainColValid, 0, validLen);
       }
-      // data is null
+      if (data != null) {
+        long dataLen = data.length;
+        mainDataDevBuff = DeviceMemoryBuffer.allocate(dataLen);
+        mainDataDevBuff.copyFromHostBuffer(data, 0, dataLen);
+      }
       if (mainColOffsets != null) {
-        long offsetsLen = OFFSET_SIZE * (mainColRows + 1);
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/ColumnVector.java:3380]: Can we please revert this change?  When copying data from the host to the device this was here to truncate the buffer, if needed, and it is now lost.

Also while you are at it could we add in a comment about why we are doing it this way.

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java [hunk 18]
```diff
@@ -596,10 +618,10 @@ long getStartStringOffset(long index) {
   }
 
   /**
-   * Get the ending byte offset for the string at index.
+   * Get the ending byte offset for the nested type or string at index.
    */
   long getEndStringOffset(long index) {
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/HostColumnVector.java:623]: Can we deprecate this API and have it call into a new one that is named more correctly?, same with getStartStringOffset?

Or revert the comment change above.  We need the assert and the comment to be consistent, and the functions between each other to be consistent.

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java [hunk 19]
```diff
@@ -1196,7 +1225,7 @@ public static HostColumnVector timestampNanoSecondsFromBoxedLongs(Long... values
   /**
    * WARNING: Debug only method to print a passed in buffer
    */
-  private static void printBuffer(HostMemoryBuffer buffer) {
+  public static void printBuffer(HostMemoryBuffer buffer) {
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/HostColumnVector.java:1228]: This API need to move to be a part of HostMemoryBuffer.

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java [hunk 20]
```diff
@@ -1276,12 +1305,20 @@ public ColumnBuilder appendLists(List... inputLists) {
...
+      return this;
+    }
+
+    private void initAndResizeOffsetBuffer(int currentIndex) {
       if (this.offsets == null) {
         offsets = HostMemoryBuffer.allocate(INIT_OFFSET_SIZE);
         offsets.setInt(0, 0);
       } else {
-        if (offsets.length <= currentIndex * OFFSET_SIZE + OFFSET_SIZE) {
+        if (offsets.length <= (currentIndex) * OFFSET_SIZE + OFFSET_SIZE) {
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/HostColumnVector.java:1321]: nit: totally unnecessary change.

File: java/src/main/native/src/ColumnVectorJni.cpp [hunk 21]
```diff
@@ -1246,12 +1248,22 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_makeCudfColumnView(
...
                                       j_null_count, 0, {offsets_column, data_column}));
     } else if (n_type == cudf::type_id::LIST) {
-         JNI_NULL_CHECK(env, j_offset, "offset is null", 0);
-         cudf::column_view *child_view = reinterpret_cast<cudf::column_view *>(j_child);
-         cudf::size_type *offsets = reinterpret_cast<cudf::size_type *>(j_offset);
-         cudf::column_view offsets_column(cudf::data_type{cudf::type_id::INT32}, size + 1, offsets);
-         ret.reset(new cudf::column_view(cudf::data_type{cudf::type_id::LIST}, size, nullptr, valid,
+      JNI_NULL_CHECK(env, j_offset, "offset is null", 0);
+      cudf::jni::native_jpointerArray<cudf::column_view> children(env, j_children);
+      cudf::column_view *child_view = reinterpret_cast<cudf::column_view *>(children[0]);
```
Comment (revans2) [inline java/src/main/native/src/ColumnVectorJni.cpp:1253]: There should be no reinterpret_cast needed.  The result of the array index should be the correct type already.

I noticed in the implementation of `operator[]` for `native_jpointerArray` there are no bounds checking.  We should either add it in there, or do it here, just in case there is an issue where someone passes in an empty array.

File: java/src/main/native/src/ColumnVectorJni.cpp [hunk 22]
```diff
@@ -1246,12 +1248,22 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_makeCudfColumnView(
...
+      cudf::size_type *offsets = reinterpret_cast<cudf::size_type *>(j_offset);
+      cudf::column_view offsets_column(cudf::data_type{cudf::type_id::INT32}, size + 1, offsets);
+      ret.reset(new cudf::column_view(cudf::data_type{cudf::type_id::LIST}, size, nullptr, valid,
                                                  j_null_count, 0, {offsets_column, *child_view}));
+   } else if (n_type == cudf::type_id::STRUCT) {
+     cudf::jni::native_jpointerArray<cudf::column_view> children(env, j_children);
+     std::vector<column_view> children_vector(children.size());
+     for (int i = 0; i < children.size(); i++) {
```
Comment (revans2) [inline java/src/main/native/src/ColumnVectorJni.cpp:1261]: I just realized that it might be possible to declare a struct with no children.  It would just have a length and validity. @mythrocks is this something that we support in cudf?  Spark supports it, but I had to jump through a lot of hoops to make it work and none of the input or output formats support it so it is close to useless.  I just wanted to know if we needed to work around it in the spark plugin and avoid it or if we needed to add in tests for it here in cudf?

File: java/src/main/native/src/ColumnVectorJni.cpp [hunk 23]
```diff
@@ -1246,12 +1248,22 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_makeCudfColumnView(
...
+       cudf::column_view *child_view = reinterpret_cast<cudf::column_view *>(children[i]);
```
Comment (revans2) [inline java/src/main/native/src/ColumnVectorJni.cpp:1262]: again the reinterpret cast is not needed you should be able to do this as a single line.

```
children_vector[i] = *children[i];
```

File: java/src/main/native/src/ColumnVectorJni.cpp (hunk 21 above)
Comment (kuhushukla) [reply java/src/main/native/src/ColumnVectorJni.cpp:1253]: Would this meet the empty array check  :  
```
  T *operator[](int index) const {
    if (wrapped.size() == 0) {
      return nullptr;
    }
    return data()[index];
  }

  T *&operator[](int index) {
    if (wrapped.size() == 0) {
      return nullptr;
    }
    return data()[index];
  }
```

Comment (kuhushukla) [reply java/src/main/native/src/ColumnVectorJni.cpp:1253]: I am not sure how to change the second call and also if returning nullptr is ok. Should I throw instead?

File: java/src/main/java/ai/rapids/cudf/HostColumnVector.java (hunk 18 above)
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/HostColumnVector.java:623]: I am going to revert the comment change as it is used for strings only.

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java (hunk 17 above)
Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/ColumnVector.java:3380]: My apologies. It works without this change of course and as you said, is needed. I have a couple things from my testing that I need to do a pass through over and this is certainly one of those that I missed. Thanks for catching that.

File: java/src/main/native/src/ColumnVectorJni.cpp (hunk 21 above)
Comment (revans2) [reply java/src/main/native/src/ColumnVectorJni.cpp:1253]: You would want to copy what the regular `native_jArray` does, but you would have to go into wrapped to get access to env. so it is not super simple.  It is probably better to have a follow on issue to fix it properly and for now just do the check in your code. 

```
  N_TYPE operator[](int index) const {
    if (orig == NULL) {
      throw_java_exception(env, NPE_CLASS, "pointer is NULL");
    }
    if (index < 0 || index >= len) {
      throw_java_exception(env, INDEX_OOB_CLASS, "NOT IN BOUNDS");
    }
    return data()[index];
  }

  N_TYPE &operator[](int index) {
    if (orig == NULL) {
      throw_java_exception(env, NPE_CLASS, "pointer is NULL");
    }
    if (index < 0 || index >= len) {
      throw_java_exception(env, INDEX_OOB_CLASS, "NOT IN BOUNDS");
    }
    return data()[index];
  }
```

Comment (kuhushukla) [reply java/src/main/native/src/ColumnVectorJni.cpp:1253]: Thanks a lot Bobby. This is informative. I will add a follow on issue, but if I am getting the fix right: I am doing this now : 
```
} else if (n_type == cudf::type_id::LIST) {
      JNI_NULL_CHECK(env, j_offset, "offset is null", 0);
      cudf::jni::native_jpointerArray<cudf::column_view> children(env, j_children);
      JNI_ARG_CHECK(env, (children.size() != 0), "children size is 0", 0);
      cudf::size_type *offsets = reinterpret_cast<cudf::size_type *>(j_offset);
      cudf::column_view offsets_column(cudf::data_type{cudf::type_id::INT32}, size + 1, offsets);
      ret.reset(new cudf::column_view(cudf::data_type{cudf::type_id::LIST}, size, nullptr, valid,
                                                 j_null_count, 0, {offsets_column, *children[0]}));
   }
```
Does this suffice or does 
```
cudf::jni::native_jpointerArray<cudf::column_view> children(env, j_children);
```
yield wrong result when size = 0? Pardon my naivety, JNI is not my forte.

File: java/src/main/native/src/ColumnVectorJni.cpp (hunk 22 above)
Comment (kuhushukla) [reply java/src/main/native/src/ColumnVectorJni.cpp:1261]: Mithun mentioned that in cudf it is possible to have 0 children structs. I will add tests here in a bit.

File: java/src/main/native/src/ColumnVectorJni.cpp (hunk 21 above)
Comment (revans2) [reply java/src/main/native/src/ColumnVectorJni.cpp:1253]: That looks fine.

File: java/src/main/native/src/ColumnVectorJni.cpp (hunk 22 above)
Comment (kuhushukla) [reply java/src/main/native/src/ColumnVectorJni.cpp:1261]: Keeping this as is.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java [hunk 24]
```diff
@@ -2956,4 +2959,80 @@ void testContiguousSplitConstructor() {
...
+      assertFalse(intChildCol.isNull(0));
+      assertFalse(intChildCol.isNull(1));
+      assertTrue(intChildCol.isNull(2));
+      assertFalse(intChildCol.isNull(3));
+      assertFalse(longChildCol.isNull(0));
+      assertFalse(longChildCol.isNull(1));
+      assertTrue(longChildCol.isNull(2));
+      assertTrue(longChildCol.isNull(3));
+      assertStructColumnsAreEqual(expected, columnVector, type);
+    }
```
Comment (kuhushukla) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:3036]: @revans2 , this is a test for children hostCV validity vector, and the code change that allows this behavior is 
[here](https://github.com/rapidsai/cudf/pull/6171/files#diff-7515abd63357ad5368785ee947cb0823R1347), appreciate any comments on the correctness, I do think it is doing what column factory on cpp side does, but I dont know why just making a column_view of type struct doesnt do the magic for us without having to do this.

Comment (revans2) [reply java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:3036]: I thought the magic happened when you created a struct column, not a column_view.  Or else creating a column_view is going to be very expensive, which it is not supposed to be.

Comment (kuhushukla) [reply java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:3036]: Agree, that seems to be the case, I just wanted to make sure I understood this correctly.

Comment (kuhushukla) [reply java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:3036]: @revans2 , I think I have covered most of the concerns, but happy to get ur reviews on any remaining lacunae.

**Comment by revans2** [comment]
I'll try to dig into it more deeply tomorrow.  But for now it is looking good.  My main asks would be to remove the WIP, and to run the full set of plugin tests against this version of cudf too.

**Comment by kuhushukla** [comment]
Thanks @revans2 ,  I will do plugin side testing next and post results tomorrow.

**Comment by kuhushukla** [comment]
Update. Took me a bit longer than I wanted but I ran the plugin with this change in cudf-0.16-snapshot jar installed and `mvn clean verify`, some udf tests and cache tests seem to fail but for unrelated reason AFAICT. I am going to try and make sure they all work and if not the failures are expected and seen even with cudf -0.16 main. Other key tests are passing.

**Comment by kuhushukla** [comment]
Verified that the the cache test failure and the udf test failure occur for my env even for nightly published cudf-0.16 version. I think we can say that this change does not break plugin to the best of my knowledge. I will follow upw tih plugin errors separately as I dont think they need to be discussed/tracked here.

**Comment by kuhushukla** [comment]
@revans2 this is ready for another review round based on my pass, but happy to address any lacunae you find. I hope to get to them today/tomorrow :) I have filed https://github.com/rapidsai/cudf/issues/6266 for the bound checks for native_jpointerArray issue.

**Review by revans2 — APPROVED** [review]
